File: desktop/src/features/communities/communityName.test.mjs
import assert from "node:assert/strict";
import test from "node:test";
import {
  reconcileCommunityName,
  withLocalCommunityName,
} from "./communityName.ts";
import { deriveCommunityName, initFirstCommunity } from "./communityStorage.ts";
const community = {
  id: "one",
  name: "buzz",
  relayUrl: "wss://buzz.example.com",
  addedAt: "2026-09-16",
};

test("canonical name replaces the display while retaining the original label", () => {
  const updated = reconcileCommunityName(community, { name: "Shared name" });
  assert.equal(updated.name, "Shared name");
  assert.equal(updated.fallbackName, "buzz");
  assert.equal(
    reconcileCommunityName(updated, { name: "Renamed" }).name,
    "Renamed",
  );
  assert.equal(
    reconcileCommunityName(updated, { name: "Shared name" }),
    updated,
  );
  assert.equal(reconcileCommunityName(updated, { name: null }).name, "buzz");
});
test("an explicit local nickname survives relay renames and can be cleared", () => {
  const named = reconcileCommunityName(community, { name: "Shared name" });
  const nicknamed = withLocalCommunityName(named, "  My nickname  ");
  const renamed = reconcileCommunityName(nicknamed, { name: "Renamed" });
  assert.equal(renamed.name, "My nickname");
  assert.equal(renamed.canonicalName, "Renamed");
  assert.equal(withLocalCommunityName(renamed, "").name, "Renamed");
});
test("private IP fallback identifies the whole address", () => {
  assert.equal(
    deriveCommunityName("ws://100.64.1.2:3000"),
    "Community (100.64.1.2)",
  );
});

test("legacy custom labels become explicit local nicknames", () => {
  const updated = reconcileCommunityName(
    { ...community, name: "My old label" },
    { name: "Shared" },
  );
  assert.equal(updated.name, "My old label");
  assert.equal(updated.localName, "My old label");
  assert.equal(updated.canonicalName, "Shared");
  assert.equal(withLocalCommunityName(updated, "").name, "Shared");
  const ip = reconcileCommunityName(
    { ...community, relayUrl: "ws://100.64.1.2:3000", name: "100" },
    { name: "Shared" },
  );
  assert.equal(ip.name, "Shared");
});

test("new invited connections follow canonical names rather than infer aliases", (t) => {
  const previousWindow = globalThis.window;
  const previousStorage = globalThis.localStorage;
  const values = new Map();
  const storage = {
    getItem: (key) => values.get(key) ?? null,
    setItem: (key, value) => values.set(key, value),
    removeItem: (key) => values.delete(key),
  };
  globalThis.window = { localStorage: storage };
  globalThis.localStorage = storage;
  t.after(() => {
    globalThis.window = previousWindow;
    globalThis.localStorage = previousStorage;
  });
  const first = initFirstCommunity(
    community.relayUrl,
    "identity",
    "Invite label",
  );
  assert.ok(first);
  assert.equal(
    reconcileCommunityName(first, { name: "Shared" }).name,
    "Shared",
  );
});

File: mobile/test/shared/community/community_name_test.dart
import 'dart:async';
import 'package:buzz/shared/community/community.dart';
import 'package:buzz/shared/community/community_name.dart';
import 'package:buzz/shared/community/community_provider.dart';
import 'package:buzz/shared/community/community_storage.dart';
import 'package:buzz/shared/community/community_icon_provider.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:hooks_riverpod/hooks_riverpod.dart';
import 'package:http/http.dart' as http;
import 'package:http/testing.dart';
import 'community_storage_test.dart';

void main() {
  final community = Community(
    id: 'one',
    name: 'buzz',
    relayUrl: 'wss://buzz.example.com',
    addedAt: DateTime(2026),
  );
  test('canonical name persists and previous label survives fallback', () {
    final named = reconcileCommunityName(
      community,
      const CommunityProfile('ATL BitLab'),
    );
    expect(named.name, 'ATL BitLab');
    expect(named.fallbackName, 'buzz');
    expect(Community.fromJson(named.toJson()).name, 'ATL BitLab');
    expect(
      reconcileCommunityName(named, const CommunityProfile(null)).name,
      'buzz',
    );
    final fresh = Community.create(
      name: 'Invite label',
      relayUrl: community.relayUrl,
    );
    expect(
      reconcileCommunityName(fresh, const CommunityProfile('Shared')).name,
      'Shared',
    );
    final legacy = reconcileCommunityName(
      community.copyWith(name: 'My label'),
      const CommunityProfile('Shared'),
    );
    expect(legacy.name, 'My label');
    expect(legacy.localName, 'My label');
    final nicknamed = named.copyWith(localName: 'Mine');
    expect(
      reconcileCommunityName(nicknamed, const CommunityProfile('New')).name,
      'Mine',
    );
    expect(
      Community.nameFromUrl('ws://100.64.1.2:3000'),
      'Community (100.64.1.2)',
    );
  });
  test(
    'read path distinguishes legacy, unnamed, malformed, and network errors',
    () async {
      var body = '{"community_profile":{"name":"ATL BitLab"}}';
      var status = 200;
      final container = ProviderContainer(
        overrides: [
          communityIconHttpClientProvider.overrideWithValue(
            MockClient((request) async {
              expect(request.url.toString(), 'https://buzz.example.com');
              expect(request.headers['Accept'], 'application/nostr+json');
              return http.Response(body, status);
            }),
          ),
        ],
      );
      addTearDown(container.dispose);
      final fetch = container.read(communityProfileFetcherProvider);
      expect((await fetch(community.relayUrl))?.name, 'ATL BitLab');
      body = '{"name":"Buzz Relay"}';
      expect(await fetch(community.relayUrl), isNull);
      body = '{"community_profile":{"name":null}}';
      expect(await fetch(community.relayUrl), isA<CommunityProfile>());
      body = '{"community_profile":{"name":32}}';
      await expectLater(fetch(community.relayUrl), throwsFormatException);
      status = 503;
      await expectLater(fetch(community.relayUrl), throwsStateError);
    },
  );
  test(
    'refresh caches names, survives offline, and drops superseded reads',
    () async {
      final storage = CommunityStorage(secure: FakeSecureStorage());
      await storage.save(community);
      var response = Future<CommunityProfile?>.value(
        const CommunityProfile('Shared'),
      );
      final container = ProviderContainer(
        overrides: [
          communityStorageProvider.overrideWithValue(storage),
          communitySnapshotWriterProvider.overrideWithValue((_) async {}),
          communityProfileFetcherProvider.overrideWithValue((_) => response),
        ],
      );
      addTearDown(container.dispose);
      await container.read(communityListProvider.future);
      final notifier = container.read(communityListProvider.notifier);
      await notifier.refreshCommunityNames();
      expect((await storage.loadAll()).single.name, 'Shared');
      response = Future.error(StateError('offline'));
      await notifier.refreshCommunityNames();
      expect(
        container.read(communityListProvider).value!.single.name,
        'Shared',
      );
      final stale = Completer<CommunityProfile?>();
      response = stale.future;
      final first = notifier.refreshCommunityNames();
      response = Future.value(const CommunityProfile('Newest'));
      await notifier.refreshCommunityNames();
      stale.complete(const CommunityProfile('Stale'));
      await first;
      expect((await storage.loadAll()).single.name, 'Newest');
      // A completing public read must preserve a concurrent nickname and
      // credentials change, including the persisted snapshot.
      final pending = Completer<CommunityProfile?>();
      response = pending.future;
      final refresh = notifier.refreshCommunityNames();
      await notifier.renameCommunity(community.id, 'My nickname');
      await notifier.addCommunity(community.copyWith(pubkey: 'new-identity'));
      pending.complete(const CommunityProfile('Renamed again'));
      await refresh;
      final stored = (await storage.loadAll()).single;
      expect(stored.name, 'My nickname');
      expect(stored.canonicalName, 'Renamed again');
      expect(stored.pubkey, 'new-identity');
    },
  );
}
